actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -54,13 +54,16 @@ def callable_cache_key?
       def collection_by_cache_keys
         seed = callable_cache_key? ? @options[:cached] : ->(i) { i }
 
+        key_base = ActiveSupport::Cache::KeyMaster.new(@view.combined_fragment_cache_key(nil))
         @collection.each_with_object({}) do |item, hash|
-          hash[expanded_cache_key(seed.call(item))] = item
+          key = ActiveSupport::Cache::KeyMaster.new(key_base)
+          key << cache_fragment_name(seed.call(item))
+          hash[key] = item
```

Comment:
I would still probably want it to be the same number of lines. Something like 

```
name = cache_fragment_name(seed.call(item))
key = ActiveSupport::Cache::KeyMaster.new(key_base, name)
hash[key] = item
```

Otherwise it gets a bit hard to read

```
key = ActiveSupport::Cache::KeyMaster.new(key_base, cache_fragment_name(seed.call(item)))
hash[key] = item
```

If you don't mind option 2 let me know. I'll need to modify the method signature to take a secondary argument. I don't want to use `*args` as I don't want to allocate an extra array per invocation.

Reply:
Not golfing for length of method here, but for coherence and readability. This method isn't concerned with how to construct a key; `expanded_cache_key` is.